pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
 	c.queue.Add(podKey)
 }
 
+func (c *Controller) checkDeletedSvc(obj interface{}) (*corev1.Service, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
+	}
+	svc, ok := deletedState.Obj.(*corev1.Service)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Service: %v", deletedState.Obj)
+	}
+	return svc, nil
+}
+
+func (c *Controller) EnqueueSvcUpdate(oldObj, newObj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both set of pods (corresponding to old and new selector) need to be considered
+	newSvc, isSvc := newObj.(*corev1.Service)
+	if !isSvc {
+		var err error
+		newSvc, err = c.checkDeletedSvc(newObj)
+		if err != nil {
+			klog.Errorf("Got error while processing event update: %v", err)
+			return
+		}
+	}
+
+	oldSvc := oldObj.(*corev1.Service)
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion {
+		return
+	}
+
+	podKeys := append(c.getPodsFromService(oldSvc), c.getPodsFromService(newSvc)...)
+	for _, podKey := range podKeys {
+		c.queue.Add(podKey)
+	}
+}
+
+func (c *Controller) EnqueueSvc(obj interface{}) {
+	svc, isSvc := obj.(*corev1.Service)
+	if !isSvc {
+		var err error
+		svc, err = c.checkDeletedSvc(obj)
+		if err != nil {
+			klog.Errorf("Got error while processing event update: %v", err)
+			return
+		}
+	}
+
+	for _, podKey := range c.getPodsFromService(svc) {
+		c.queue.Add(podKey)
+	}
+}
+
+func (c *Controller) getPodsFromService(svc *corev1.Service) []string {
+	var pods []string
+	annotations := svc.GetAnnotations()
+	if val, ok := annotations[NPLServiceAnnotation]; !ok || (ok && val != "true") {
+		return pods
+	}
+
+	appSelectorValue, ok := svc.Spec.Selector["app"]
+	if !ok {
+		return pods
+	}
+
+	listOptions := metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{"app": appSelectorValue}).String(),
+		FieldSelector: fields.OneTermEqualSelector("spec.nodeName", c.NodeName).String(),
+	}
+	podList, err := c.kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), listOptions)
```

Comment:
This won't scall well. We should use the PodLister of PodInformer to avoid more queries.